syncer/syncer_test.go

```diff
@@ -1074,53 +1084,63 @@ func (s *testSyncerSuite) TestRun(c *C) {
 
 	go syncer.Process(ctx, resultCh)
 
-	testCases1 := []struct {
-		sql      string
-		tp       opType
-		sqlInJob string
-		arg      interface{}
-	}{
+	sqls1 := []string{
+		"create database if not exists test_1",
+		"create table if not exists test_1.t_1(id int primary key, name varchar(24))",
+		"create table if not exists test_1.t_2(id int primary key, name varchar(24))",
+		"insert into test_1.t_1 values(1, 'a')",
+		"alter table test_1.t_1 add index index1(name)",
+		"insert into test_1.t_1 values(2, 'b')",
+		"delete from test_1.t_1 where id = 1",
+		"update test_1.t_1 set id = 1 where id = 2", // will find casuality and then generate flush job
+	}
+	expectJobs1 := []*expectJob{
 		{
-			"create database if not exists test_1",
 			ddl,
 			"CREATE DATABASE IF NOT EXISTS `test_1`",
 			nil,
 		}, {
-			"create table if not exists test_1.t_1(id int)",
 			ddl,
-			"CREATE TABLE IF NOT EXISTS `test_1`.`t_1` (`id` INT)",
+			"CREATE TABLE IF NOT EXISTS `test_1`.`t_1` (`id` INT PRIMARY KEY,`name` VARCHAR(24))",
 			nil,
 		}, {
-			"create table if not exists test_1.t_2(id int)",
 			ddl,
-			"CREATE TABLE IF NOT EXISTS `test_1`.`t_2` (`id` INT)",
+			"CREATE TABLE IF NOT EXISTS `test_1`.`t_2` (`id` INT PRIMARY KEY,`name` VARCHAR(24))",
 			nil,
 		}, {
-			"insert into test_1.t_1 values(1)",
 			insert,
-			"REPLACE INTO `test_1`.`t_1` (`id`) VALUES (?);",
-			int64(580981944116838401),
+			"REPLACE INTO `test_1`.`t_1` (`id`,`name`) VALUES (?,?);",
+			[]interface{}{int64(580981944116838401), "a"},
 		}, {
-			"alter table test_1.t_1 add index index1(id)",
 			ddl,
-			"ALTER TABLE `test_1`.`t_1` ADD INDEX `index1`(`id`)",
+			"ALTER TABLE `test_1`.`t_1` ADD INDEX `index1`(`name`)",
 			nil,
 		}, {
-			"insert into test_1.t_1 values(2)",
 			insert,
-			"REPLACE INTO `test_1`.`t_1` (`id`) VALUES (?);",
-			int64(580981944116838402),
+			"REPLACE INTO `test_1`.`t_1` (`id`,`name`) VALUES (?,?);",
+			[]interface{}{int64(580981944116838402), "b"},
 		}, {
-			"delete from test_1.t_1 where id = 1",
 			del,
 			"DELETE FROM `test_1`.`t_1` WHERE `id` = ? LIMIT 1;",
-			int64(580981944116838401),
+			[]interface{}{int64(580981944116838401)},
+		}, {
+			flush,
+			"",
+			nil,
+		}, {
+			update,
+			"DELETE FROM `test_1`.`t_1` WHERE `id` = ? LIMIT 1;",
```

Comment:
safe mode?

Reply:
yes, I add a comment `in first 5 minutes, safe mode is true, will split update to delete + replace`